### DISCUSSION TITLE
Question : Is to possible to read  the Message field in human readable format inside the circuit

### DISCUSSION THREAD

**Discussion by vishal-kanna** [Consensys/gnark#1176] [General]
I was working with Ethan Buchman e snark account https://github.com/informalsystems/cosmos-sdk/blob/gnark50/crypto/keys/gnark/eddsa/eddsa.go .I want to read the message in side the circuit .
Is it possible?

**Comment by ivokub** [comment]
Can you elaborate what do you mean by reading the message? Are you interested in debug printing the intermediate values?

For this, we have `api.Debug(v.. frontend.Variable)` which allows to print the values at solving time, or when run in a test engine. I could have more feedback on how to insert and use it when you post fully running example.

**Comment by vishal-kanna** [comment]
@ivokub Hey,Thanks for the reply.
Here reading the message mean printing the `circuit.Message` value.When i try to print the value using `api.Println()` I could print the value but not in human readable format .Here it is `DBG multisig1.go:68 > cricuit.Message>>>>>>. 19920802116712217054180446958248493647949177859458659037637035837968196228499`
For example : `circuit.Message= "something" `
I want to print that message value("something") inside the circuit logic.Is it possible to print?

**Comment by ivokub** [comment]
All the variables inside the circuit are actually field elements (essentially integers modulo scalar field), so the human-readable representation for it is actually an integer.

There is a natural way to encode messages as integers though -> you look at every letter as a message as byte and then represent these bytes as base-256 encoded integer. Similarly, you can have the encoding the other way, but keep in mind that for every byte there may not be an printable letter (particularly for byte values [127,255]).

**Comment by vishal-kanna** [comment]
Thank you @ivokub 
Is type assertion possible inside the circuit logic.
Example: Converting the fronted.Variable to Big.Int or any other type ?

**Comment by ivokub** [comment]
> Thank you @ivokub Is type assertion possible inside the circuit logic. Example: Converting the fronted.Variable to Big.Int or any other type ?

By default it is not possible. The circuit as defined is turned into a list of constraints depending on the arithmetization (R1CS/Plonk) and as such we cannot really have conversions as the actual values are only assigned at proving time and we don't have the circuit code anymore. When using `api.Println` then we add some additional debug information to the compiled circuit which marks which element we wanted to print and it is printed at proving/solving time.

However, there is a tool for debugging circuits. There is [`test.IsSolved`](https://pkg.go.dev/github.com/consensys/gnark@v0.10.0/test#IsSolved) method which instead of compiling and proving, just runs all the computation on `*big.Int`. When you run it, then you should be able to perform type assertion
```go
a := api.Mul(x, y)
aa, ok := a.(*big.Int)
fmt.Println(aa.String())
```
but we always don't work with `*big.Int` type, but use occasionally uint, uint64 etc. types. So your type switch should be bigger a la:
```go
a := api.Mul(x, y)
switch aa := a.(type) {
case *big.Int:
/// print big.Int
case big.Int:
/// print big.Int
case uint64:
/// print uint64
... // etc
}
```

**Comment by vishal-kanna** [comment]
Thankyou @ivokub
